keps/sig-apps/0008-topology-aware-workload-controllers.md

```diff
@@ -0,0 +1,350 @@
+  controllers updates, scale up/down.
+
+### Non-Goals
+
+Other topology related design discussions, such as services and volumes, are out
+of scope. 
+
+## Background
+
+Here is how pods placement (scheduling) and deletion is managed today:
+
+| Controller  | Creation time                   | Scale up            | Scale down                          | Rolling update                               |
+|-------------|---------------------------------|---------------------|-------------------------------------|----------------------------------------------|
+| DaemonSet   | DaemonSet or scheduler          | -                   | -                                   | DaemonSet (kills pods by health)             |
+| StatefulSet | Scheduler                       | Scheduler           | StatefulSet (kills pods by ordinal) | StatefulSet (kills pods by ordinal)          |
+| ReplicaSet  | Scheduler                       | Scheduler           | ReplicaSet (kills pod by health)    | -                                            |
+| Deployment  | ReplicaSet creation or scale up | ReplicaSet scale up | Replicasets scale down              | ReplicaSet scale up + ReplicaSets scale down |
+
+### Spreading 
+
+First, DaemonSet is special, because it is the only workload controller whose
+pods don’t go through default scheduler. In the future, DaemonSet may create
+pods that get scheduled by the default scheduler. Either way, DaemonSet pods
+spread equally, because a DaemonSet schedules one pod per node. 
+
+For other workload controllers, during pods placement (pods creation), the
+default scheduler offers best-effort spreading (`SelectorSpreadPriority`) and
+explicit spreading rule (pod anti-affinity). 
+
+However, when scaling down, the controller decides which pods to kill first,
+without going through default scheduler. Therefore, scaling down ignores
+scheduling policy. 
+
+Controller rolling update is done by killing old replicas and creating new ones.
+Similar to scaling down cases, when old replicas are brought down, the decision
+doesn’t go through the default scheduler, either. On the other hand, the
+creation of new replicas goes through the default scheduler, so the spreading
+only works after the rolling update is done. 
+
+### Update Patterns 
+
+The workload controllers that support rolling update feature offer customizable
+update strategies. Users can specify the number of old replicas to bring down or
+the number of new replicas to create during a rollout. 
+
+Existing update patterns are not topology-aware. For example, users cannot
+specify things like: 
+
+* Update by failure domain: Update one failure domain at a time (staged domain)
+* Update across failure domains: Only allow X number of replicas to be down per
+  failure domain 
+
+## Proposal
+
+We can think of updates as planned disruptions. When a workload controller
+performs a rolling update, it will evict (instead of deleting) old pods and
```

Comment:
With deletion you can delete a Pod on a Node that's down - it sets the deletionTimestamp and continue. Will eviction work even in case the node is down and in NotReady state?

Reply:
evict is delete plus the pdb checks so yes it should work